### ISSUE TITLE
Ktor instrumentation not producing correctly nesting spans

### ISSUE THREAD

**Issue by eivbsmed** [open-telemetry/opentelemetry-java-instrumentation#6776]
Hi,

I'm trying to get instrumentation working in Ktor as it's increasing in use in or org.
We are using the latest version of the Java agent to do auto instrumentation.
The Server engine is set to Netty and the Ktor http clinet is set to use OkHttp as the engine.

OkHttp is picked up by the Auto instrumentation and looks ok.

After finding there was a Ktor instrumentation this is how the app is set up:

```
fun main(args: Array<String>): Unit = io.ktor.server.netty.EngineMain.main(args)

//Get Open telemetry from Auto instrumentation
val openTelemetry = GlobalOpenTelemetry.get()

@Suppress("unused")
fun Application.module() {
    
    install(KtorServerTracing) {
        setOpenTelemetry(openTelemetry)
    }
```
This all looks to work ok

Traces are collected and have the right structure. 1 Http call in and 2 out in the right structure

The problem shows up when we are trying to add a span to a function in between.
The span then shows one call in and 3 parallel ones
Code:
```
    suspend fun get(id: ID): Item {
        val span = openT.spanBuilder("getOrganization").startSpan()
        span.addEvent("Hey!")
        val jointResult = withContext(span.asContextElement()) {
            val response = client.get("https://example.com/$is")
            val response2 = client.get("https://example2.com/$is")
            val result: EnhetsregisterUnit = response.deserialize()
            combine(response,response2
        }
        span.end()
        return jointResult
    }
```

Result
```
O---*
      | span 1 function
      | get 1
      | get 2
```

Wanted result
```
O---*
      | fun 1 ---*
                    | get 1
                    | get 2
```

To organise it we also tried @WithSpan

Code:
```
    @WithSpan
    suspend fun get(id: ID): Item {
        Span.current().addEvent("Hey!")
        val response = client.get("https://example.com/$is")
        val response2 = client.get("https://example2.com/$is")
        val result: EnhetsregisterUnit = response.deserialize()
        combine(response,response2
        return jointResult
    }
```

The result of this code was a bit better but the @WithSpan made 3 spans that make it looks like the function was called 3 times.  
```
O---*
      | fun 1 ---*
                    | get 1
      | fun 1 ---*
                    | get 2
      | fun 1 (2)
```

Wanted result
```
O---*
      | fun 1 ---*
                    | get 1
                    | get 2
```

Using versions 1.18.0 and 1.18.0 alpha on the Kotlin instrumentation

**Environment**
Eclipse Adoptium OpenJDK 64-Bit Server VM 17.0.1+12
Linux 5.19.0-76051900-generic

We are current using Jearger to look at the traces.

**Comment by laurit** [comment]
`@WithSpan` currently doesn't work correctly with coroutines because it tracks each method call, but suspend functions may be called multiple times.
As for the incorrect nesting I guess it is because for coroutines context is kept in coroutine context (the `withContext(span.asContextElement())`) but okhttp instrumentation reads context from the regular `ThreadLocal` based context. You could try something like
```
coroutineContext.getOpenTelemetryContext().makeCurrent().use { 
  client.get("https://example.com/$is")
}
```
and see if this changes anything (I don't know much kotlin so so sorry if it doesn't work at all). Though I believe this would be ok only if you know that `client.get` doesn't suspend as otherwise you could have the exit happen on a different thread from the enter and end up with a mess.

**Comment by eivbsmed** [comment]
It's the Ktor http client. So the get has this signature `public suspend inline fun HttpClient.get(`

I tried the suggested code and it did not change anything, i still got the flat trace where the function call and all the gets where on the same level.

**Comment by eivbsmed** [comment]
Looks like i have i always have to make sure i always have use `withContext(span.asContextElement()){...<stuff>..}` and then when creating a sub span set  `Span.spanBuilder(name).setParent(coroutineContext.getOpenTelemetryContext())` as the parrent when creating a new span. So it's not automatically supporting coroutines in the instrumentation.

So that will correctly nest my spans. But i have still not found out how to fix the auto instrumentation of the OkHttp library to use the span from the closes coroutinesContext as it parrent so they are still wrong. If i can't find a way I'll probably look int writing a plugin for the Ktor http client to do it. 

As for fixing the span tree in or own code i found this from Oliver https://slack-chats.kotlinlang.org/t/488833/and-no-need-for-sir-i-m-just-oliver-laughing and adapted it from to set the parent from the context.

Trying to figure out how to use context  in a good way as Im new to Kotlin my self.  

This is currently the code i have adjusted from Oliver, putting it here if others come looking:
```kotlin
import io.opentelemetry.api.common.AttributeKey
import io.opentelemetry.api.common.Attributes
import io.opentelemetry.api.trace.Span
import io.opentelemetry.api.trace.SpanBuilder
import io.opentelemetry.api.trace.StatusCode
import io.opentelemetry.api.trace.Tracer
import io.opentelemetry.extension.kotlin.asContextElement
import io.opentelemetry.extension.kotlin.getOpenTelemetryContext
import kotlinx.coroutines.CoroutineName
import kotlinx.coroutines.withContext
import kotlin.coroutines.cancellation.CancellationException
import kotlin.coroutines.coroutineContext

val EXCEPTION_TYPE = AttributeKey.stringKey("exception.type")!!
val EXCEPTION_MESSAGE = AttributeKey.stringKey("exception.type")!!


context(Tracer)
suspend fun <Result> withSpan(
    name: String,
    parameters: (SpanBuilder.() -> Unit)? = null,
    exceptionIsError: (Throwable) -> Boolean = { it !is CancellationException },
    block: suspend (span: Span?) -> Result
): Result {
    val span: Span = spanBuilder(name).run {
        setParent(coroutineContext.getOpenTelemetryContext())
        if (parameters != null) {
            parameters()
        }
        coroutineContext[CoroutineName]?.let {
            setAttribute("coroutine.name", it.name)
        }

        startSpan()
    }

    return withContext(span.asContextElement()) {
        try {
            block(span).also {
                span.setStatus(StatusCode.OK)
            }
        } catch (throwable: Throwable) {
            if (exceptionIsError(throwable)) {
                span.setStatus(StatusCode.ERROR)
                span.recordException(throwable)
            } else {

                val attributes = with(Attributes.builder()){
                    put(EXCEPTION_TYPE,throwable.javaClass.name)
                    put(EXCEPTION_MESSAGE, (throwable.message ?: "(none)"))
                    build()
                }

                span.addEvent(
                    "Completed with exception",
                    attributes
                )
                span.setStatus(StatusCode.OK)
            }
            throw throwable
        } finally {
            span.end()
        }
    }
}
```

This let me write some thing like this:

I'll probably end up whit renaming `whitSpan` to just `span` and look more into how i can use the context receiver better. If it should be in the context of a trace or a opentelemetry instance

```kotlin
with(trace) {
            withSpan("Duh") {
                Thread.sleep(300)
            }
            withSpan("Class.Function") {
                   Thread.sleep(300)
                withSpan("Class.Function.1") {
                    Thread.sleep(300)
                }
                withSpan(""Class.Function.4") {
                    Thread.sleep(300)
                }
            }
        }
```

But still stuck on getting the OkHttp auto instrumented spans right place in the trace as i still get them parented to the root span. Any ideas and pointers welcome

**Comment by mpeyper** [comment]
Hey @eivbsmed, did you ever get a solution for the `OkHttp` spans appearing in the right place?  I'm facing similar issues for the same reasons in #7124 and hoping someone has worked this out.

**Comment by trask** [comment]
I believe this is resolved by #9149. You can test either with the [latest snapshot release](https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/main/CONTRIBUTING.md#snapshot-builds) or with the upcoming 1.29.0 release in 2 days.
